providers/implementations/keymgmt/hybrid_kmgmt.c

```diff
@@ -0,0 +1,782 @@
+    unsigned char **pptr = NULL;
+    size_t *plen = NULL;
+    OSSL_PARAM *lprm = NULL;
+    unsigned int i, j, k, nparams;
+    size_t data_size, t;
+    int res = 0;
+
+    nparams = ossl_param_nelem(params);
+    if (nparams == 0)
+        return 0;
+
+    lprm = OPENSSL_memdup(params, (1 + nparams) * sizeof(*params));
+    pptr = OPENSSL_malloc(nparams * sizeof(*pptr));
+    plen = OPENSSL_malloc(nparams * sizeof(*plen));
+    if (lprm == NULL || pptr == NULL || plen == NULL)
+        goto err;
+
+    for (j = 0; j < nparams; j++) {
+        pptr[j] = params[j].data;
+        plen[j] = params[j].data_size;
+    }
+
+    for (i = 0; i < key->info->num_algs; i++) {
+        for (k = j = 0; j < nparams; j++) {
+            /*
+             * Variable length unsigned integer encoding.
+             *
+             * The top two bits of the first byte specify the length of the
+             * size field.  The number of _additional_ length bytes is:
+             * first_byte >> 6.  The bottom six bits of the first byte
+             * and all of the subsequent bytes represent an unsigned integer
+             * in big endian format.  The maximum length is thus 1Gb - 1 bytes.
+             * A length of zero indicates that this parameter is not applicable
+             * for this algorithm.
+             *
+             * The length is followed by a data block of the specified size.
+             * This is treated as opaque data for the sub-algorithm.
+             *
+             * The rationale behind this encoding is that we do not know
+             * what parameters (or their associated sizes) are required for
+             * each sub-algorithm.
+             */
+            if (plen[j] == 0)
+                goto err;
+            t = (*pptr[j] & 0300) >> 6;
+            data_size = *pptr[j]++ & 077;
+            if (--plen[j] < t)
+                goto err;
+            plen[j] -= t;
+            while (t-- > 0)
+                data_size = data_size << 8 | *pptr[j]++;
+
+            /* Only include this param if it's got an associated value */
+            if (data_size > 0) {
+                if (plen[j] < data_size)
+                    goto err;
+                lprm[k] = params[j];
+                lprm[k].data = pptr[j];
+                lprm[k++].data_size = data_size;
+                pptr[j] += data_size;
+                plen[j] -= data_size;
+            }
+        }
+        /* Don't bother if there are no params */
+        if (k > 0) {
+            lprm[k] = OSSL_PARAM_construct_end();
+            ctx = EVP_PKEY_CTX_new_from_name(key->libctx,
+                                             key->info->alg[i].name,
+                                             key->propq);
+            if (ctx == NULL
+                    || EVP_PKEY_fromdata_init(ctx) <= 0
+                    || EVP_PKEY_fromdata(ctx, key->keys + i, selection,
+                                         lprm) <= 0)
+                goto err;
+        }
+    }
```

Comment:
I wonder if this is the right way to do it, i.e. to join the secret and public key values of these keys into one param.

Should we instead just suffix (or prefix) the param names and merge the arrays? I.e., have `alg1_priv` and `alg2_priv` params in the resulting OSSL_PARAM array?

Reply:
I started doing that and it resulted in messier code.  This is clean and simple in comparison.

Reply:
I disagree. Maybe my proposal would require more code but I do not think the performance is a concern here as a few strncmps should be negligible in comparison to the computations of the encaps/decaps operations.

Reply:
Having tried both approaches, I disagree.

You are correct, performance isn't critical but it never hurts.